libs/vkd3d/swapchain.c

```diff
@@ -2526,6 +2596,111 @@ static void dxgi_vk_swap_chain_update_frame_statistics(struct dxgi_vk_swap_chain
     spinlock_release(&chain->frame_statistics.lock);
 }
 
+static void dxgi_vk_swap_chain_platform_sleep_for_ns(struct platform_sleep_state *sleep_state, uint64_t duration_ns)
+{
+#if defined(_WIN32)
+    LARGE_INTEGER ticks;
+
+    if (sleep_state->NtDelayExecution)
+    {
+        /* Ticks are in units of 100ns, with negative values indicating a
+         * duration and positive values an absolute time in tick counts. */
+        ticks.QuadPart = (int64_t)duration_ns / -100;
+
+        /* Set alertable = false to avoid wineserver round-trips */
+        sleep_state->NtDelayExecution(FALSE, &ticks);
+    }
+    else
+    {
+        Sleep(duration_ns / 1000000);
+    }
+#else
+    struct timespec duration;
+
+    (void)sleep_state;
+
+    duration.tv_sec = duration_ns / 1000000000;
+    duration.tv_nsec = duration_ns % 1000000000;
+
+    /* ignore interrupts */
+    nanosleep(&duration, NULL);
+#endif
+}
+
+static void dxgi_vk_swap_chain_delay_next_frame(struct dxgi_vk_swap_chain *chain, uint64_t current_time_ns)
+{
+    struct platform_sleep_state *sleep_state = &chain->frame_rate_limit.sleep_state;
+    uint64_t window_start_time_ns, window_total_time_ns;
+    uint64_t sleep_threshold_ns, sleep_duration_ns;
+    uint64_t next_deadline_ns = 0;
+
+    pthread_mutex_lock(&chain->frame_rate_limit.lock);
+
+    if (chain->frame_rate_limit.target_interval_ns)
+    {
+        if (!chain->frame_rate_limit.enable)
+        {
+            const uint32_t window_size = ARRAY_SIZE(chain->frame_rate_limit.frame_timestamps);
+
+            window_start_time_ns = chain->frame_rate_limit.frame_timestamps[chain->frame_rate_limit.frame_window_index % window_size];
+
+            chain->frame_rate_limit.frame_timestamps[chain->frame_rate_limit.frame_window_index % window_size] = current_time_ns;
+            chain->frame_rate_limit.frame_window_index += 1;
+
+            window_total_time_ns = current_time_ns - window_start_time_ns;
+
+            /* Enable the frame rate limiter if frames are being delivered faster than expected.
+             * Discard the first window worth of data since it may be inaccurate, especially if
+             * we do not hit the present_wait path. */
+            if (chain->frame_rate_limit.frame_window_index >= 2 * window_size)
+            {
+                chain->frame_rate_limit.enable = 103 * window_total_time_ns < 100 * window_size * chain->frame_rate_limit.target_interval_ns;
+
+                if (chain->frame_rate_limit.enable)
```

Comment:
Do we want to engage this if vsync is off? Or will DXVK set target frame rate to 0.0 then?

Reply:
DXVK handles the vsync off case.